File: Difficulty/Medium/[61] Rotate List/solution.cpp
/*
 * @lc app=leetcode id=61 lang=cpp
 *
 * [61] Rotate List
 */

//  * Definition for singly-linked list.
// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode(int x) : val(x), next(NULL) {}
// };

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr) return nullptr;
        if (k == 0 || head->next == nullptr) return head;
        ListNode* pre = head, *lst = head, *Trv = head;
        int length = 1;
        while (Trv->next != nullptr){
            Trv = Trv->next;
            length++;
        }
        int rotate_len = k % length;
        if (rotate_len == 0) return head;
        int pre_i = 0, lst_i = 0;
        while(lst->next != nullptr){
            lst = lst->next;
            if (lst_i - pre_i == rotate_len){
                pre = pre->next;
                pre_i++;
            }
            lst_i++;
        }
        ListNode* new_head = pre->next;
        pre->next = nullptr;
        lst->next = head;
        return new_head;
    }
};

File: Difficulty/Medium/[61] Rotate List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return
        if k == 0 or head.next == None:
            return head
        pre, lst, trv, length = head, head, head, 1
        while trv.next != None:
            trv = trv.next
            length += 1
        rotate_len = k % length
        if rotate_len == 0:
            return head
        pre_i, lst_i = 0, 0
        while lst.next != None:
            lst = lst.next
            if lst_i - pre_i == rotate_len:
                pre = pre.next
                pre_i += 1
            lst_i += 1
        new_head = pre.next
        pre.next = None
        lst.next = head
        return new_head
